packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6688,11 +6688,16 @@ abstract class RenderObjectElement extends Element {
   }
 
   void _updateParentData(ParentDataWidget<ParentData> parentDataWidget) {
-    bool applyParentData = true;
-    assert(() {
-      try {
-        if (!parentDataWidget.debugIsValidRenderObject(renderObject)) {
-          applyParentData = false;
+    if (parentDataWidget.debugIsValidRenderObject(renderObject)) {
```

Comment:
This should probably not named started with debug, and should probably be convert into a private static method

Reply:
I agree. It affects the control flow, and should not be exclusive to assert mode. But it was written this way ever since the first day, and not only `debugIsValidRenderObject`, but also the method that it depends on, `debugTypicalAncestorWidgetClass`, are prefixed by `debug`, the latter being intended for overriding and is indeed overridden by many classes. We might want to ask Ian (the first author) whether it was intentional.

Reply:
We shouldn't use debugTypicalAncestorWidgetClass in release mode, because it'll force widgets to be included that might otherwise get optimised out. But I think we can still throw in release mode even if the error message isn't as clear. We have lots of precedent for that.

Reply:
(debugIsValidRenderObject should be renamed though)

Reply:
I've renamed `debugIsValidRenderObject` to make it static and private.

Are you saying that incorrect usage of ParentDataWidget should still throw and cause a crash to users in release mode? I'd have imagined that release mode should be as fault-tolerant as possible, and attempt to deliver a good experience even if it doesn't follow Flutter's rules. In my experience using e.g. `Flexible` outside a `Column` does not cause any visual/usability issues, only a broken tree which is flagged in the logs in debug mode. 

A similar case to this that current exists is using a `Positioned` outside a `Stack`. There is no point in using it outside a `Stack`, and it is an error, but it does not currently break the release build and gives an error in the console in debug mode.